### PR TITLE
set interface mtu based on lwaftr v4/v6 mtu

### PR DESCRIPTION
**Pull request by mwiget**
Fix for https://github.com/Igalia/snabb/issues/487

It calculates the MTU for the interface based on the ipv4_mtu and ipv6_mtu, by taking the numerically larger one and add 14 to cover the size of the ethernet header. The change will also print out the mtu to STDOUT together with the interface name.

Fixing README files for snabbvmx ipv4_interface and ipv6_interface settings.

### CODE REVIEW

**Comment by chrgraf** [comment]
Thanks marcel. One addition Please. If mtu is not configured under soft wire it shall be derived from interface mtu please.

Thanks

Christian

Von meinem Windows Phone gesendet

---

Von: Marcel Wigetmailto:notifications@github.com
Gesendet: ‎15.‎10.‎2016 07:25
An: Igalia/snabbmailto:snabb@noreply.github.com
Betreff: [Igalia/snabb] set interface mtu based on lwaftr v4/v6 mtu (#488)

Fix for #487https://github.com/Igalia/snabb/issues/487

It calculates the MTU for the interface based on the ipv4_mtu and ipv6_mtu, by taking the numerically larger one and add 14 to cover the size of the ethernet header. The change will also print out the mtu to STDOUT together with the interface name.

Fixing README files for snabbvmx ipv4_interface and ipv6_interface settings.

---

You can view, comment on, or merge this pull request online at:

  https://github.com/Igalia/snabb/pull/488

Commit Summary
-   set interface mtu based on lwaftr v4/v6 mtu

File Changes
-   M src/program/snabbvmx/README.mdhttps://github.com/Igalia/snabb/pull/488/files#diff-0 (11)
-   M src/program/snabbvmx/lwaftr/READMEhttps://github.com/Igalia/snabb/pull/488/files#diff-1 (6)
-   M src/program/snabbvmx/lwaftr/lwaftr.luahttps://github.com/Igalia/snabb/pull/488/files#diff-2 (9)
-   M src/program/snabbvmx/lwaftr/setup.luahttps://github.com/Igalia/snabb/pull/488/files#diff-3 (10)

Patch Links:
-   https://github.com/Igalia/snabb/pull/488.patch
-   https://github.com/Igalia/snabb/pull/488.diff

**Comment by mwiget** [comment]
@chrgraf: There is no mtu other than the tunnel-payload-mtu and tunnel-path-mru, both are mandatory. If no br-instance is configured, then I set the MTU to DEFAULT_MTU = 9500, which is used in transparent mode, so all packets will be sent to the vMX 0/0/x interface, where the MTU from the vMX will be applied.

**Comment by chrgraf** [comment]
Hi Marcel,

I have meant the [edit interfaces] mtu or even better by issuing a operational command show interface ge000 | match mtu.

Having two places to configure mtu (edit interface and edit softwire] does create unnecessary conflicts.

Ideally if the [edit softwire] has no mtu configured it would be good if it would be derived from the ge000 mtu.

Many thanks

Christian

Von meinem Windows Phone gesendet

---

Von: Marcel Wigetmailto:notifications@github.com
Gesendet: ‎15.‎10.‎2016 09:54
An: Igalia/snabbmailto:snabb@noreply.github.com
Cc: Christian Grafmailto:cgraf@juniper.net; Mentionmailto:mention@noreply.github.com
Betreff: Re: [Igalia/snabb] set interface mtu based on lwaftr v4/v6 mtu (#488)

@chrgrafhttps://github.com/chrgraf: There is no mtu other than the tunnel-payload-mtu and tunnel-path-mru, both are mandatory. If no br-instance is configured, then I set the MTU to DEFAULT_MTU = 9500, which is used in transparent mode, so all packets will be sent to the vMX 0/0/x interface, where the MTU from the vMX will be applied.

**Comment by mwiget** [comment]
@chrgraf: Don't want to mix the configs of the 'micro VNF' snabbvmx with the config of the vMX. Remember that subscriber packets never reach the vMX, so the MTU must be configured by the br-instance. I also can't make the [xe|ge]-0/0/x mtu the only MTU, because the ietf-softwire I-D made the mtu and mru settings mandatory. In addition, Junos allows setting the IFD and IFF MTU seperately, which will make it complicated to derive it.

Bottom line, IMHO the straight forward method is to derive the interface (L2) MTU for snabb 10GE from the softwire mtu/mru settings.

File: src/program/snabbvmx/lwaftr/lwaftr.lua [hunk 1]
```diff
@@ -141,11 +141,18 @@ function run(args)
 
    local vlan = conf.settings and conf.settings.vlan or false
 
+   local mtu = DEFAULT_MTU
+   if lwconf.ipv6_mtu then
+     mtu = lwconf.ipv6_mtu + 14
```
Comment (dpino) [inline src/program/snabbvmx/lwaftr/lwaftr.lua]: Better to use constant `ethernet_header_size` in `apps/lwaftr/constants.lua`. Also if vlan is true, should increase `mtu` by 4.

File: src/program/snabbvmx/lwaftr/setup.lua [hunk 2]
```diff
@@ -48,20 +48,24 @@ local function load_virt (c, nic_id, lwconf, interface)
    }
 
    local v4_nic_name, v6_nic_name = nic_id..'_v4', nic_id..'v6'
+   local v4_mtu = lwconf.ipv4_mtu + 14
```
Comment (dpino) [inline src/program/snabbvmx/lwaftr/setup.lua]: Similar to the previous case.

**Comment by chrgraf** [comment]
Hi Marcel,

Thanks for this explanation - its the way to go.

Christian

---

Christian Graf
Senior Consulting Engineer
Telephone +49 5651/336174
Mobile +49 175/2929 167
cgraf@juniper.net
www.juniper.net

Juniper Networks GmbH
Sitz der Gesellschaft / Registered office: Oskar-Schlemmer-Str. 15, 80807 München
Registergericht / Commercial registry: München, HRB 205571
Geschäftsführer / Directors: Johannes Albers, Stephen Byrne, Terrance Floyd Spidell

---

This email contains material that is confidential. The content of this email is for the sole use of the intended recipient(s).
Any review or distribution by persons other than the intended recipient(s) without the express permission of Juniper Networks is strictly prohibited.
If you are not the intended recipient, please contact the sender and delete/destroy all copies of this email and any related attachments.
Juniper does not guarantee the accuracy or completeness of third party materials or information

**Comment by mwiget** [comment]
@dpino thanks for the suggestion. Updated the files and also covered the places I missed earlier. We seem to have too many cases for loading phy and virt and soon one more when adding virtio_net. Refactoring would be a good idea at some point ;-)

**Review by dpino — APPROVED** [review]
